### ISSUE TITLE
Laravel Passport JWT integration

### ISSUE THREAD

**Issue by nikolaynesov** [websanova/vue-auth#185]
HI,

I checked the https://github.com/websanova/vue-auth/issues/23 issue, but didn't find an answer to my question there.

I am using` Laravel + Passport` and `JWT` to consume my API With JS.
Passport middleware attaches the laravel_token cookie to my outgoing responses successfully. 
If I use `vue-resource` requests - I can see `laravel_token` token transferred correctly and recognized be the app.

When  I do :

```
this.$auth.login({
                    body: this.form.data,
                    params: {},
                    success: function () {},
                    error: function () {},
                    rememberMe:  this.form.data.remember,
                    redirect: {name: 'dashboard'},
                    fetchUser: true
                });
```
Fetching the user does not work as it does not transfer the `laravel_token` cookie to server (Getting Unauthorized error). At the same time when I use just `vue-resource` (after this.$auth.login ) separately - everything works fine and `laravel_token` is transferred.

This is how I use it with Vue:

```
Vue.use(require('@websanova/vue-auth'), {
    auth:   require('@websanova/vue-auth/drivers/auth/bearer.js'),
    http:   require('@websanova/vue-auth/drivers/http/vue-resource.1.x.js'),
    router: require('@websanova/vue-auth/drivers/router/vue-router.2.x.js'),
});
```

Can you please tell me how to make it work correctly with Laravel Passwort JWT? maybe I missed something in docs/auth driver?

**Comment by websanova** [comment]
You need to write a custom driver as the currently available ones only work
with header.

Just console .log the current  driver, you will receive it's just couple
lines of code ;)

**Comment by nikolaynesov** [comment]
Sorry for asking a lot of questions here, but I've got another one. 
From what I found in testing server app, for plugin to successfully fetch user and set authenticated status to `true`, the response structure should be:

```
'status' => 'success',
'data'=> $user
```

I am using https://github.com/thephpleague/fractal and DataArraySerializer serializer, so I don't have 'status', I understand the status by http response code. Looks like I need to override either Fractal Serializer or Fetch function in plugin right?

Are you going to make any https://github.com/vuejs/vuex integration? Mutators + Actions looks useful. What if I am fetching the user with few dependencies and also want a some point later reload not the whole user instance but just one dependency (balance or address for ex.)? Can I set user balance from JS and track this parameter changes?

Thank you for your responses and this cool plugin.

**Comment by websanova** [comment]
Ok, so

*First:*

The status doesn't matter at all, it just looking for the `data` which you
can change using `parseUserData`.

https://github.com/websanova/vue-auth#parseuserdata-_parseuserdata

*Second:*

I don't use vuex, but you can use `$auth.user()` directly anywhere in your
templates, it's already a binded Vue object.

If you like you can do whatever with your vuex by setting it $auth.user()
either directly or using `Object.assign()` (I would assume). To update the
user data in vue-auth you can use `$auth.user(obj)`.  And to update it with
new from server you can use `$auth.fetch()`.

https://github.com/websanova/vue-auth#user
https://github.com/websanova/vue-auth#fetch

**Comment by heisjoe** [comment]
@nikolaynesov 
i am using vue-auth with laravel and laravel use JWT package too.
my custom driver is
```
module.exports = {

    request: function (req, token) {
        this.options.http._setHeaders.call(this, req, {Authorization: 'Bearer ' + token});
    },

    response: function (res) {
        return res.data.token;
    }
};

```
it is work now .
fetch users function my laravel code is :

```
$return = [
            'status' => 'success',
            'data' => User::UserInfo($user->id)
        ];
        return response()->json($return);
```
Hi: @websanova 
I meet one question , with vue-auth  login function 

```
this.$auth.login({
    data: this.formDatas,
    rememberMe: false,
    fetchUser: true,
    success: (res) => {
        this.loading = false;
    },
    error: (err) => {
        console.log(err);
    }
});

```

when i use email and error password, i can not fetch error message in error function , it seems like
browser intercept.  @nikolaynesov  do you get this question?

**Comment by websanova** [comment]
Well, what's the response code from the server?

**Comment by nikolaynesov** [comment]
@heisjoe I am not using JWT package, I am using Laravel Passport. It has a specific middleware that sets `laravel_token` cookie. I don't use local storage in browser.
So with this approach I don't need to store JWT key in plugin, I just need to get the auth user.
Currently I am getting the user with default :

```
parseUserData (data) {

        console.log(data.data);
        return data.data;
    }
```
Console shows me the user data correctly, I see something like this:

![screen shot 2017-07-03 at 14 05 43](https://user-images.githubusercontent.com/17743574/27790346-bf9ca5b2-5ff8-11e7-816c-ffd8ea709cd2.png)

but, `this.$auth.user()` is always an empty object. `this.$auth.ready()` is `true` and `this.$auth.check()` is `false`...

@websanova what I am missing?

**Comment by websanova** [comment]
Now you asking me totally different question.

And this is in the docs brw.

**Comment by nikolaynesov** [comment]
@websanova yep, I managed to pass the `laravel_token` cookie to login request, but now the issue is that I am getting the user, `parseUserData` logs the user object, but `this.$auth.user()` is always empty. According to docs https://github.com/websanova/vue-auth#user-data 

> The plugin is designed to expect the user object to be in the data parameter of the response.

And it is there so I expect `this.$auth.user()` to have user object after parsing, not an empty object right?

**Comment by heisjoe** [comment]
@nikolaynesov 

ok , i know, at first i did it at that time ,  vue-auth + laravel + passport. and it is inappropriate

> Vue-auth supports simple token based and JSON Web Tokens (JWT) authentication

so, you should use vue-auth with laravel jwt package, it is my suggest

**Comment by websanova** [comment]
Right, so you are not using $auth.ready() correctly I would presume.

It should be used on elements only, it's just returning a reactive
variable, it's not a callback.

https://github.com/websanova/vue-auth#ready

So if you try to run $auth.ready() in an if it will give you the value at
that time only.

**Comment by nikolaynesov** [comment]
@websanova @heisjoe ,

Finally, we decided that Passport is an overhead in our case and shifted to `tymondesigns/jwt-auth` safely. So `websanova/vue-auth` works perfect now 👍  But I would not recommend using it with Passport. It is possible but there is too much headache. Also, for those who found this issue - make sure that you really need whole Passport thing for just a JWT authentication because in most cases when you don't need `oauth`, `tymondesigns/jwt-auth `implementation is just enough. 

@websanova, thank you for the package and your patience answering the questions! Closing the issue.
